File: two-number-sum/main.two.ts
// O(n log(n)) time | O(1) space

const twoNumberSum = (array: number[], targetSum: number) => {
	array.sort((a, b) => a - b);
	let leftIdx = 0;
	let rightIdx = array.length - 1;

	while (leftIdx < rightIdx) {
		const currentSum = array[leftIdx] + array[rightIdx];
		if (currentSum === targetSum) {
			return [array[leftIdx], array[rightIdx]];
		} else if (currentSum < targetSum) {
			leftIdx++;
		} else if (currentSum > targetSum) {
			rightIdx--;
		}
	}

	return [];
};

export {};

File: two-number-sum/main.two.cpp
#include <vector>
using namespace std;

vector<int> twoNumberSum(vector<int> array, int targetSum) {
  int leftIdx = 0;
	int rightIdx = array.size() - 1;
	sort(array.begin(), array.end());
  
	while (leftIdx < rightIdx) {
		int currentSum = array[leftIdx] + array[rightIdx];
		
		if (currentSum == targetSum) {
			return {array[leftIdx], array[rightIdx]};
		} else if (currentSum < targetSum) {
			leftIdx++;
		} else if (currentSum > targetSum) {
			rightIdx--;
		}
	}
	return {};
}
